awx/main/migrations/_inventory_source.py

```diff
@@ -62,3 +65,84 @@ def remove_azure_inventory_sources(apps, schema_editor):
     logger.debug("Removing all Azure InventorySource from database.")
     InventorySource.objects.filter(source='azure').delete()
 
+
+def _get_instance_id(from_dict, new_id, default=''):
+    '''logic mostly duplicated with inventory_import command Command._get_instance_id
+    frozen in time here, for purposes of migrations
+    '''
+    instance_id = default
+    for key in new_id.split('.'):
+        if not hasattr(from_dict, 'get'):
+            instance_id = default
+            break
+        instance_id = from_dict.get(key, default)
+        from_dict = instance_id
+    return smart_text(instance_id)
+
+
+def _get_instance_id_for_upgrade(host, new_id):
+    if host.instance_id:
+        # this should not have happened, but nothing to really do about it...
+        logger.debug('Unexpectedly, host {}-{} has instance_id set'.format(host.name, host.pk))
+        return None
+    host_vars = parse_yaml_or_json(host.variables)
```

Comment:
For large inventories, I _could_ see this being a potentially time consuming thing if you've got a lost of hostvars.  This is potentially tens of thousands of "parse a large-ish JSON document, compare things, save a new value to the database."

Have you done any timing of this with large, fake datasets (i.e., a few thousand hosts w/ really large JSON blobs).

Reply:
This is the reason I like the reverse migration.

https://gist.github.com/AlanCoding/40a9209dac5cc065c65faa673229127b

I can dive into speculation with a few specifics from this. I have 166 inventories with the same gce source imported redundantly. If you do the division, that's a dozen hosts that are imported from this import.

Either we are worried about large single gce sources, or we are worried about a large number of gce sources. In the case of the former, we already do parsing of the JSON for all hosts in the inventory import, so as long as we're not _worse_ than that, doesn't really matter. The former is more murky. Sure 166x12 isn't huge, we might be more worried about 30x5,000. If the concern is migration time, the current data seems relatively reassuring. Memory could be a genuine problem if some things do not get garbage collected.

For the size of the variables... these tend to be 1,000 characters (as the JSON formatted string) before the inventory plugin, and 4,000 characters after. We will only be migrating the pre-plugin versions of these hosts (except for a very small population of AWX users).